File: src/cli/commands/mining.js
// Mining Commands
const { UI, COLORS, ICONS } = require("../../util/UI.js");
const { shortenAddress } = require("../../util/AddressHelper.js");

function mineCommand(vorpal, walletManager, blockchain, p2p) {
  vorpal
    .command(
      "mine <wallet>",
      "Mine block. <wallet> can be wallet name or address",
    )
    .alias("m")
    .action(function (args, callback) {
      try {
        let minerAddress;
        let displayName = args.wallet;

        // 1. Thử tìm trong local wallets
        try {
          minerAddress = walletManager.getPublicKey(args.wallet);
        } catch {
          // 2. Nếu là địa chỉ (64 hex chars) -> dùng trực tiếp
          if (args.wallet.length === 64 && /^[a-f0-9]+$/i.test(args.wallet)) {
            minerAddress = args.wallet;
            displayName = shortenAddress(args.wallet);
          } else {
            // 3. Tìm prefix trong blockchain
            const allBalances = blockchain.getAllBalances();
            const matches = Object.keys(allBalances).filter((addr) => {
              if (typeof addr !== "string" || addr.length !== 64) return false;
              return addr.toLowerCase().startsWith(args.wallet.toLowerCase());
            });

            if (matches.length === 1) {
              minerAddress = matches[0];
              displayName = shortenAddress(minerAddress);
            } else if (matches.length > 1) {
              throw new Error(`Multiple addresses match "${args.wallet}"`);
            } else {
              throw new Error(`Wallet/address not found: "${args.wallet}"`);
            }
          }
        }

        this.log(UI.info(`Mining block for ${displayName}...`));

        const newBlock = blockchain.mineBlock(minerAddress);

        if (newBlock) {
          const content = [
            UI.keyValue(
              "Block",
              `#${COLORS.cyan}${newBlock.index}${COLORS.reset}`,
            ),
            UI.keyValue(
              "Transactions",
              `${COLORS.yellow}${newBlock.transactions.length}${COLORS.reset}`,
            ),
            UI.keyValue(
              "Nonce",
              `${COLORS.dim}${newBlock.nonce}${COLORS.reset}`,
            ),
            UI.keyValue(
              "Hash",
              `${COLORS.dim}${newBlock.hash.slice(0, 16)}...${COLORS.reset}`,
            ),
          ].join("\n");

          this.log("\n" + UI.box(content, `${ICONS.mining} Block Mined`, 45));

          const reward = newBlock.coinbaseTx ? newBlock.coinbaseTx.amount : 0;
          this.log(UI.success(`Reward: ${reward} coins`));

          // Broadcast block
          p2p.broadcastNewBlock(newBlock);
          this.log(UI.info(`Broadcasted to ${p2p.peers.length} peer(s)\n`));
        } else {
          this.log(UI.error("Mining failed"));
        }
      } catch (err) {
        this.log(UI.error(`Error: ${err.message}`));
      }
      callback();
    });
}

// Store auto-mine interval globally
let autoMineInterval = null;

function autoMineCommand(vorpal, walletManager, blockchain, p2p) {
  vorpal
    .command(
      "automine <wallet> [interval]",
      "Auto-mine when mempool has transactions. Interval in seconds (default: 10)",
    )
    .alias("am")
    .action(function (args, callback) {
      try {
        // Stop existing auto-mine
        if (autoMineInterval) {
          clearInterval(autoMineInterval);
          autoMineInterval = null;
          this.log(UI.info("Auto-mine stopped"));
          callback();
          return;
        }

        let minerAddress;
        try {
          minerAddress = walletManager.getPublicKey(args.wallet);
        } catch {
          if (args.wallet.length === 64 && /^[a-f0-9]+$/i.test(args.wallet)) {
            minerAddress = args.wallet;
          } else {
            throw new Error(`Wallet not found: ${args.wallet}`);
          }
        }

        const intervalSec = parseInt(args.interval) || 10;
        const displayName = shortenAddress(minerAddress);

        this.log(
          UI.success(
            `Auto-mine started for ${displayName} (every ${intervalSec}s)`,
          ),
        );
        this.log(UI.info("Run 'automine' again to stop\n"));

        const self = this;

        autoMineInterval = setInterval(() => {
          if (blockchain.mempool.length > 0) {
            self.log(
              UI.info(
                `Auto-mining ${blockchain.mempool.length} pending tx(s)...`,
              ),
            );
            try {
              const newBlock = blockchain.mineBlock(minerAddress);
              if (newBlock) {
                const reward = newBlock.coinbaseTx
                  ? newBlock.coinbaseTx.amount
                  : 0;
                self.log(
                  UI.success(
                    `Block #${newBlock.index} mined! Reward: ${reward} coins`,
                  ),
                );
                p2p.broadcastNewBlock(newBlock);
              }
            } catch (err) {
              self.log(UI.error(`Auto-mine error: ${err.message}`));
            }
          }
        }, intervalSec * 1000);
      } catch (err) {
        this.log(UI.error(`Error: ${err.message}`));
      }
      callback();
    });
}

function stopAutoMineCommand(vorpal) {
  vorpal
    .command("stopautomine", "Stop auto-mining")
    .alias("sam")
    .action(function (args, callback) {
      if (autoMineInterval) {
        clearInterval(autoMineInterval);
        autoMineInterval = null;
        this.log(UI.success("Auto-mine stopped"));
      } else {
        this.log(UI.info("Auto-mine is not running"));
      }
      callback();
    });
}

module.exports = {
  mineCommand,
  autoMineCommand,
  stopAutoMineCommand,
};

File: src/cli/mining.rs
//! CLI — lệnh mining: mine, automine, stopautomine.

use std::sync::Arc;

use crate::node::NodeHandle;
use crate::p2p::{messages, P2P};
use crate::util::{self, *};

use super::resolve_address;

pub(super) async fn cmd_mine(node: &NodeHandle, p2p: &Arc<P2P>, wallet: &str) {
    let miner_address = match resolve_address(node, wallet).await {
        Ok(a) => a,
        Err(e) => {
            println!("{}", util::error(&e));
            return;
        }
    };
    let display = util::shorten_address(&miner_address);
    println!("{}", util::info(&format!("Mining block for {display}...")));

    // PoW chạy ngoài lock — node vẫn phản hồi các lệnh khác trong lúc mine
    let Some(block) = crate::node::mine_block(node, &miner_address).await else {
        println!(
            "{}",
            util::warning("Chain changed while mining — transactions returned to mempool")
        );
        return;
    };

    let lines = vec![
        util::key_value("Block", &format!("#{CYAN}{}{RESET}", block.index)),
        util::key_value(
            "Transactions",
            &format!("{YELLOW}{}{RESET}", block.transactions.len()),
        ),
        util::key_value("Nonce", &format!("{DIM}{}{RESET}", block.nonce)),
        util::key_value(
            "Hash",
            &format!("{DIM}{}{RESET}", util::prefix(&block.hash, 16)),
        ),
    ];
    println!("\n{}", util::box_lines(&lines, "⛏ Block Mined", 45));
    let reward = block.coinbase_tx.as_ref().map(|c| c.amount).unwrap_or(0);
    println!(
        "{}",
        util::success(&format!("Reward: {} coins", util::fmt_micro(reward)))
    );

    p2p.broadcast(&messages::new_block(&block));
    println!(
        "{}\n",
        util::info(&format!("Broadcasted to {} peer(s)", p2p.get_peers().len()))
    );
}

pub(super) async fn cmd_automine(node: &NodeHandle, p2p: &Arc<P2P>, args: &[&str]) {
    // Toggle: đang chạy -> stop (thả lock trước khi stop lock lại)
    if node.lock().await.auto_mine.is_some() {
        cmd_stop_automine(node).await;
        return;
    }

    let mut n = node.lock().await;

    let Some(wallet) = args.first() else {
        println!("{}", util::error("Usage: automine <wallet> [interval]"));
        return;
    };
    let miner_address = match n.wallets.get_address(wallet) {
        Ok(a) => a,
        Err(_) if util::is_valid_address(wallet) => wallet.to_string(),
        Err(_) => {
            println!("{}", util::error(&format!("Wallet not found: {wallet}")));
            return;
        }
    };
    let interval = args
        .get(1)
        .and_then(|s| s.parse::<u64>().ok())
        .unwrap_or(10);
    let display = util::shorten_address(&miner_address);
    println!(
        "{}",
        util::success(&format!(
            "Auto-mine started for {display} (every {interval}s)"
        ))
    );
    println!("{}\n", util::info("Run 'automine' again to stop"));

    n.auto_mine = Some((miner_address, interval));
    let node_clone = node.clone();
    let p2p_clone = p2p.clone();
    n.auto_mine_task = Some(tokio::spawn(auto_mine_task(node_clone, p2p_clone)));
}

async fn auto_mine_task(node: NodeHandle, p2p: Arc<P2P>) {
    loop {
        let interval = {
            let n = node.lock().await;
            let Some((_, interval)) = n.auto_mine.clone() else {
                break;
            };
            interval
        };
        tokio::time::sleep(std::time::Duration::from_secs(interval)).await;

        let block = {
            let n = node.lock().await;
            if n.auto_mine.is_none() {
                break;
            }
            if n.blockchain.mempool.is_empty() {
                continue;
            }
            println!(
                "{}",
                util::info(&format!(
                    "Auto-mining {} pending tx(s)...",
                    n.blockchain.mempool.len()
                ))
            );
            let miner = n.auto_mine.as_ref().unwrap().0.clone();
            drop(n);
            match crate::node::mine_block(&node, &miner).await {
                Some(b) => b,
                None => continue, // chain đổi giữa chừng — vòng sau thử lại
            }
        };
        let reward = block.coinbase_tx.as_ref().map(|c| c.amount).unwrap_or(0);
        println!(
            "{}",
            util::success(&format!(
                "Block #{} mined! Reward: {} coins",
                block.index,
                util::fmt_micro(reward)
            ))
        );
        p2p.broadcast(&messages::new_block(&block));
    }
}

pub(super) async fn cmd_stop_automine(node: &NodeHandle) {
    let mut n = node.lock().await;
    if let Some(task) = n.auto_mine_task.take() {
        task.abort();
        n.auto_mine = None;
        println!("{}", util::success("Auto-mine stopped"));
    } else {
        println!("{}", util::info("Auto-mine is not running"));
    }
}
